Sources/NIO/CircularBuffer.swift

```diff
@@ -97,17 +174,37 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
         self.buffer = newBacking
     }
 
-    /// Remove the front element of the ring buffer.
+    /// Removes & returns the item at `position` from the buffer
     ///
-    /// *O(1)*
-    public mutating func removeFirst() -> E {
-        guard let value = self.buffer[self.headIdx] else {
-            preconditionFailure("CircularBuffer is empty")
+    /// - Parameter position: The index of the item to be removed from the buffer.
+    ///
+    /// *O(1)* if the position is `headIdx` or `tailIdx`.
+    /// otherwise
+    /// *O(n)* where *n* is the number of elements between `position` and `tailIdx`.
+    public mutating func remove(at position: Int) -> E {
+        precondition(self.indices.contains(position), "Position out of bounds.")
+        var bufferIndex = self.bufferIndex(ofIndex: position)
+        let element = self.buffer[bufferIndex]!
+
+        switch bufferIndex {
+        case self.headIdx:
+            self.headIdx = (self.headIdx + 1) & self.mask
```

Comment:
Now you have `bufferIndex(after:)` you can rewrite this to use `self.headIdx = self.bufferIndex(after: self.headIdx)`. It reads a bit more nicely. :)

Reply:
Nice 👍